runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/getdata/ApplianceGetDataClient.java

```diff
@@ -0,0 +1,253 @@
+import org.apache.beam.runners.dataflow.worker.windmill.ApplianceWindmillClient;
+import org.apache.beam.runners.dataflow.worker.windmill.Windmill;
+import org.apache.beam.runners.dataflow.worker.windmill.Windmill.ComputationGetDataRequest;
+import org.apache.beam.runners.dataflow.worker.windmill.work.refresh.HeartbeatSender;
+import org.apache.beam.runners.dataflow.worker.windmill.work.refresh.Heartbeats;
+import org.apache.beam.sdk.annotations.Internal;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.base.Preconditions;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.Iterables;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.util.concurrent.SettableFuture;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/** Appliance implementation of {@link GetDataClient}. */
+@Internal
+@ThreadSafe
+public final class ApplianceGetDataClient implements GetDataClient, WorkRefreshClient {
+  private static final int MAX_READS_PER_BATCH = 60;
+  private static final int MAX_ACTIVE_READS = 10;
+
+  private final ApplianceWindmillClient windmillClient;
+  private final ThrottlingGetDataMetricTracker getDataMetricTracker;
+
+  @GuardedBy("this")
+  private final List<ReadBatch> pendingReadBatches;
+
+  @GuardedBy("this")
+  private int activeReadThreads;
+
+  public ApplianceGetDataClient(
+      ApplianceWindmillClient windmillClient, ThrottlingGetDataMetricTracker getDataMetricTracker) {
+    this.windmillClient = windmillClient;
+    this.getDataMetricTracker = getDataMetricTracker;
+    this.pendingReadBatches = new ArrayList<>();
+    this.activeReadThreads = 0;
+  }
+
+  public static GetDataClient create(
+      ApplianceWindmillClient windmillClient, ThrottlingGetDataMetricTracker getDataMetricTracker) {
+    return new ApplianceGetDataClient(windmillClient, getDataMetricTracker);
+  }
+
+  @Override
+  public Windmill.KeyedGetDataResponse getStateData(
+      String computation, Windmill.KeyedGetDataRequest request) {
+    try (AutoCloseable ignored =
+        getDataMetricTracker.trackSingleCallWithThrottling(
+            ThrottlingGetDataMetricTracker.Type.STATE)) {
+      SettableFuture<Windmill.KeyedGetDataResponse> response = SettableFuture.create();
+      ReadBatch batch = addToReadBatch(new QueueEntry(computation, request, response));
+      if (batch != null) {
+        issueReadBatch(batch);
+      }
+      return response.get();
+    } catch (Exception e) {
+      throw new GetDataException(
+          "Error occurred fetching state for computation="
+              + computation
+              + ", key="
+              + request.getShardingKey(),
+          e);
+    }
+  }
+
+  @Override
+  public Windmill.GlobalData getSideInputData(Windmill.GlobalDataRequest request) {
+    try (AutoCloseable ignored =
+        getDataMetricTracker.trackSingleCallWithThrottling(
+            ThrottlingGetDataMetricTracker.Type.STATE)) {
```

Comment:
this shoudl be SIDE_INPUT

since we just have STATE and SIDE_INPUT, I wonder if it wuold be less error prone to have the public methods on the tracker be trackGetDataWithThrottling and trackGetSideInputWithThrottling instead of the enum.

Reply:
done